src/ensembl/src/shared/utils/lruCache.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * This is a cache implementing the Least Recently Used algorithm
+ * for preventing it from growing indefinitely. It also can invalidate
+ * individual items based on the max time during which an item is allowed
+ * to be retrieved from the cache.
+ *
+ * The Least Recently Used algorithm requires coordinated use of a doubly linked list
+ * for storing cached values, and a map for quick access to the nodes of the list
+ */
+
+interface LRUCacheInterface {
+  get(key: string): any;
+  set(key: string, value: any): void;
+  reset(): void;
+}
+
+interface LinkedListNodeInterface {
+  value: any;
+  next: LinkedListNode | null;
+  previous: LinkedListNode | null;
+  timestamp: number;
+}
+
+type LRUCacheOptions = {
+  size?: number; // max number of items that can be stored in cache; defaults to 1000
+  maxAge?: number; // time period, in milliseconds, during which a cache item is retrievable
+};
+
+type LinkedListNodeOptions = {
+  key: string;
+  value: any;
+  timestamp: number;
+};
+
+const DEFAULT_SIZE = 1000;
+
+export default class LRUCache implements LRUCacheInterface {
+  private size: number;
+  private maxAge?: number;
+  private currentSize = 0;
+  private head: LinkedListNode | null;
+  private tail: LinkedListNode | null;
+  private listNodesMap: Map<string, LinkedListNode>;
+
+  constructor(options: LRUCacheOptions = {}) {
+    if (options.size && options.size <= 0) {
+      throw new Error('Cache size must be a positive integer');
+    }
+    this.size = options.size || DEFAULT_SIZE;
+    this.maxAge = options.maxAge;
+
+    this.head = null;
+    this.tail = null;
+    this.listNodesMap = new Map();
+  }
+
+  public get(key: string) {
+    const cachedNode = this.listNodesMap.get(key);
+    if (cachedNode) {
+      if (this.isExpiredNode(cachedNode)) {
+        this.removeNode(cachedNode);
+        return;
+      }
+      this.moveNodeToHead(cachedNode);
+      cachedNode.timestamp = Date.now();
+      return cachedNode.value;
+    }
+  }
+
+  public set(key: string, value: any) {
+    const savedNode = this.listNodesMap.get(key);
+    if (savedNode) {
```

Comment:
Will this ever be true?

If I understand correctly, we `set` only when `get` returns nothing.

Reply:
> If I understand correctly, we set only when get returns nothing.
That's how the api service is using the cache. But if you just take LRUCache as a standalone unit, there is nothing preventing you from calling the method `set` on it without first calling the `get`. I think it should be capable of handling this kind of misuse gracefully.